src/storage/page_buffer.c

```diff
@@ -7272,56 +7290,63 @@ pgbuf_delete_from_hash_chain (THREAD_ENTRY * thread_p, PGBUF_BCB * bufptr)
 /*
  * pgbuf_lock_page () - Puts a buffer lock on the buffer lock chain
  *   return: If success, PGBUF_LOCK_HOLDER, otherwise PGBUF_LOCK_WAITER
- *   hash_anchor(in):
  *   vpid(in):
+ *   fetch_mode (in):
+ *   hash_anchor(in):
+ *   perf (in/out): page fix performance monitoring helper
  *
  * Note: This function is invoked only when the page is not in the buffer hash
  *       chain. The caller is holding hash_anchor->hash_mutex.
  *       Before return, the thread releases hash_anchor->hash_mutex.
  */
-static int
-pgbuf_lock_page (THREAD_ENTRY * thread_p, PGBUF_BUFFER_HASH * hash_anchor, const VPID * vpid)
+static PGBUF_LOCK_MODE
+pgbuf_lock_page (THREAD_ENTRY * const thread_p, const VPID * const vpid, PAGE_FETCH_MODE fetch_mode,
+		 PGBUF_BUFFER_HASH * const hash_anchor, PGBUF_FIX_PERF * perf)
 {
+  PGBUF_LOCK_MODE pgbuf_lock_state = PGBUF_LOCK_UNKNOWN;
+
 #if defined(SERVER_MODE)
   PGBUF_BUFFER_LOCK *cur_buffer_lock;
-  THREAD_ENTRY *cur_thrd_entry;
   TSC_TICKS start_tick, end_tick;
   UINT64 lock_wait_time = 0;
 
-  /* the caller is holding hash_anchor->hash_mutex */
-  /* check whether the page is in the Buffer Lock Chain */
-
-  if (thread_p == NULL)
+  // TODO: more accurate comments
+  /* The page is not found in the hash chain the caller is holding hash_anchor->hash_mutex */
+  if (er_errid () == ER_CSS_PTHREAD_MUTEX_TRYLOCK)
     {
-      assert (thread_p != NULL);
-      thread_p = thread_get_thread_entry_info ();
+      pthread_mutex_unlock (&hash_anchor->hash_mutex);
+      PGBUF_BCB_CHECK_MUTEX_LEAKS ();
+      return PGBUF_LOCK_UNKNOWN;
     }
 
-  cur_thrd_entry = thread_p;
+  /* the caller is holding hash_anchor->hash_mutex */
+  /* check whether the page is in the Buffer Lock Chain */
+
   cur_buffer_lock = hash_anchor->lock_next;
 
-  /* find vpid in buffer lock chain */
+  /* find vpid in buffer lock chain; entry being the result of indexing via a hash, there might be vpid collisions */
   while (cur_buffer_lock != NULL)
     {
       if (VPID_EQ (&(cur_buffer_lock->vpid), vpid))
 	{
-	  /* found */
-	  cur_thrd_entry->next_wait_thrd = cur_buffer_lock->next_wait_thrd;
-	  cur_buffer_lock->next_wait_thrd = cur_thrd_entry;
-	  pgbuf_sleep (cur_thrd_entry, &hash_anchor->hash_mutex);
+	  /* found page, enlist this thread at the head of the page's buffer lock chain
```

Comment:
Nitpicking: I would move this part or some of the code in this scope to another function to improve code readability.

Reply:
Extracted two functions: pgbuf_buffer_lock_enlist_thread and pgbuf_buffer_lock_delist_thread this the scope.